File: src/CSharp/CodeCracker/Usage/JsonNetAnalyzer.cs
using System;
using System.Collections.Immutable;
using System.Reflection;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;

namespace CodeCracker.CSharp.Usage
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class JsonNetAnalyzer : DiagnosticAnalyzer
    {
        internal const string Title = "Your Json syntax is wrong";
        internal const string MessageFormat = "{0}";
        internal const string Category = SupportedCategories.Usage;
        const string Description = "This diagnostic checks the json string and triggers if the parsing fail "
            + "by throwing an exception.";

        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
            DiagnosticId.JsonNet.ToDiagnosticId(),
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Error,
            isEnabledByDefault: true,
            description: Description,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.JsonNet));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSyntaxNodeAction(c => Analyzer(c, "DeserializeObject", "Newtonsoft.Json.JsonConvert.DeserializeObject<T>(string)"), SyntaxKind.InvocationExpression);
            context.RegisterSyntaxNodeAction(c => Analyzer(c, "Parse", "Newtonsoft.Json.Linq.JObject.Parse(string)"), SyntaxKind.InvocationExpression);
            context.RegisterSyntaxNodeAction(c => Analyzer(c, "Parse", "Newtonsoft.Json.Linq.JArray.Parse(string)"), SyntaxKind.InvocationExpression);
        }

        private static void Analyzer(SyntaxNodeAnalysisContext context, string methodName, string methodFullDefinition)
        {
            if (context.IsGenerated()) return;
            var invocationExpression = (InvocationExpressionSyntax)context.Node;

            var memberExpression = invocationExpression.Expression as MemberAccessExpressionSyntax;
            if (memberExpression?.Name?.Identifier.ValueText != methodName) return;

            var memberSymbol = context.SemanticModel.GetSymbolInfo(memberExpression).Symbol;
            if (memberSymbol?.OriginalDefinition?.ToString() != methodFullDefinition) return;

            var argumentList = invocationExpression.ArgumentList;
            if ((argumentList?.Arguments.Count ?? 0) != 1) return;

            var literalParameter = argumentList.Arguments[0].Expression as LiteralExpressionSyntax;
            if (literalParameter == null) return;

            var jsonOpt = context.SemanticModel.GetConstantValue(literalParameter);
            var json = jsonOpt.Value as string;

            CheckJsonValue(context, literalParameter, json);
        }

        private static void CheckJsonValue(SyntaxNodeAnalysisContext context, LiteralExpressionSyntax literalParameter,
            string json)
        {
            try
            {
                parseMethodInfo.Value.Invoke(null, new[] { json });
            }
            catch (Exception ex)
            {
                var diag = Diagnostic.Create(Rule, literalParameter.GetLocation(), ex.InnerException.Message);
                context.ReportDiagnostic(diag);
            }
        }

        private static readonly Lazy<Type> jObjectType = new Lazy<Type>(() => Type.GetType("Newtonsoft.Json.Linq.JObject, Newtonsoft.Json"));
        private static readonly Lazy<MethodInfo> parseMethodInfo = new Lazy<MethodInfo>(() => jObjectType.Value.GetRuntimeMethod("Parse", new[] { typeof(string) }));
    }
}

File: src/VisualBasic/CodeCracker/Usage/JsonNetAnalyzer.vb
Imports System.Collections.Immutable
Imports System.Reflection
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Usage
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class JsonNetAnalyzer
        Inherits DiagnosticAnalyzer

        Friend Const Title = "Your JSON syntax is wrong."
        Friend Const MessageFormat = "{0}"
        Private Const Description = "This diagnostic checks the Json string and triggers if the parsing fails by throwning an exception"

        Friend Shared Rule As New DiagnosticDescriptor(
            DiagnosticId.JsonNet.ToDiagnosticId(),
            Title,
            MessageFormat,
            SupportedCategories.Usage,
            SeverityConfigurations.Current(DiagnosticId.JsonNet),
            isEnabledByDefault:=True,
            description:=Description,
            helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.JsonNet))

        Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
            Get
                Return ImmutableArray.Create(Rule)
            End Get
        End Property

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(Sub(c) Analyze(c, "DeserializeObject", "Public Shared Overloads Function DeserializeObject(Of T)(value As String) As T"), SyntaxKind.InvocationExpression)
            context.RegisterSyntaxNodeAction(Sub(c) Analyze(c, "Parse", "Public Shared Overloads Function Parse(json As String) As Newtonsoft.Json.Linq.JObject"), SyntaxKind.InvocationExpression)
        End Sub

        Private Sub Analyze(context As SyntaxNodeAnalysisContext, methodName As String, methodFullDefinition As String)
            If (context.IsGenerated()) Then Return
            Dim invocationExpression = DirectCast(context.Node, InvocationExpressionSyntax)
            Dim memberExpression = TryCast(invocationExpression.Expression, MemberAccessExpressionSyntax)
            If memberExpression?.Name?.Identifier.ValueText <> methodName Then Exit Sub

            Dim memberSymbol = context.SemanticModel.GetSymbolInfo(memberExpression).Symbol
            If memberSymbol?.OriginalDefinition?.ToString() <> methodFullDefinition Then Exit Sub

            Dim argumentList = invocationExpression.ArgumentList
            If If(argumentList?.Arguments.Count, 0) <> 1 Then Exit Sub

            Dim literalParameter = TryCast(argumentList.Arguments(0).GetExpression(), LiteralExpressionSyntax)
            If literalParameter Is Nothing Then Exit Sub

            Dim jsonOpt = context.SemanticModel.GetConstantValue(literalParameter)
            Dim json = jsonOpt.Value.ToString()

            CheckJsonValue(context, literalParameter, json)
        End Sub

        Private Shared Sub CheckJsonValue(context As SyntaxNodeAnalysisContext, literalParameter As LiteralExpressionSyntax, json As String)
            Try
                parseMethodInfo.Value.Invoke(Nothing, {json})
            Catch ex As Exception
                Dim diag = Diagnostic.Create(Rule, literalParameter.GetLocation(), ex.InnerException.Message)
                context.ReportDiagnostic(diag)
            End Try
        End Sub

        Private Shared ReadOnly jObjectType As New Lazy(Of Type)(Function() System.Type.GetType("Newtonsoft.Json.Linq.JObject, Newtonsoft.Json"))
        Private Shared ReadOnly parseMethodInfo As New Lazy(Of MethodInfo)(Function() jObjectType.Value.GetRuntimeMethod("Parse", {GetType(String)}))
    End Class
End Namespace
